src/replicache.test.ts

```diff
@@ -1797,3 +1808,100 @@ test('clientID', async () => {
   expect(clientID3).to.match(re);
   expect(clientID3).to.equal(clientID);
 });
+
+// Only used for type checking
+test.skip('mut [type checking only]', async () => {
+  const rep = new Replicache({
+    mut: {
+      a: (tx: WriteTransaction) => {
+        console.log(tx);
+        return 42;
+      },
+      b: (tx: WriteTransaction, x: number) => {
+        console.log(tx, x);
+        return 'hi';
+      },
+
+      // Return void
+      c: (tx: WriteTransaction) => {
+        console.log(tx);
+      },
+      d: (tx: WriteTransaction, x: number) => {
+        console.log(tx, x);
+      },
+
+      e: async (tx: WriteTransaction) => {
+        console.log(tx);
+        return 42;
+      },
+      f: async (tx: WriteTransaction, x: number) => {
+        console.log(tx, x);
+        return 'hi';
+      },
+
+      // Return void
+      g: async (tx: WriteTransaction) => {
+        console.log(tx);
+      },
+      h: async (tx: WriteTransaction, x: number) => {
+        console.log(tx, x);
+      },
+
+      // This should be flagged as an error but I need to use `any` for the arg since I need bivariance an TS uses covariant here.
+      // @ts-expect-error
+      i: (tx: WriteTransaction, d: Date) => {
```

Comment:
So if I'm understanding this correctly we got:

- no need to repeat the mutator name as a string separately from field name
- mutators are strongly typed (when I hover them at `rep.mut.foo` I will see the full signature, and if I type the wrong thing it will complain)

And the only thing we could *not* get:

- The mutators passed to `ReplicacheOptions` allow any type for their argument, when it should be restricted to JSONValue

Do I have it right? This seems pretty amaze if so. We should throw an error anyway when dev tries to pass something non-jsony through the args, so they'll notice it early.

Reply:
You have that right.

> We should throw an error anyway when dev tries to pass something non-jsony through the args, so they'll notice it early.

We can do that... let's see...

```js
function isJSON(v) {
  switch (typeof v) {
    case 'number':
    case 'string':
    case 'boolean':
      return true;
    case 'object': {
      if (v === null) {
        return true;
      }
      const proto = Object.getPrototypeOf(v);
      return proto === Object.prototype || proto === Array.prototype;
    }
    default:
      return false;   
  }
}
```

Reply:
> And the only thing we could not get:

I think the reason this worked before was that I manually extracted the `Args` part in `register`. Now I'm not that sure cannot be done because we operate on the options object...

Reply:
`isJSON` needs recursive check too... Maybe we should only do this in dev?